### PR TITLE
Add shifting for sort order

### LINKED ISSUE #4144: Drop max() in SortableModel for a database trigger instead
For concurrent access we should implement, if possible, a postgres trigger to populate the `sort_order`.

### PR DESCRIPTION
**Pull request by NyanKiyoshi**
Definition of done:
- [x] Bulk shifting is possible
- [x] Supports gaps between sort orders
- [x] Supports bulk operations properly
- [x] Proper handling of `null` sort orders
- [x] Migration makes sure sort order are unique
- [x] Tests are written
- [x] Tests for the mutations are updated
- [x] Tests for the resolvers are updated/rewritten

Partially fixes #4144.

### Pull Request Checklist

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [ ] All visible strings are translated with proper context.
1. [ ] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] Database migration files are up to date.
1. [x] The changes are tested.
1. [x] The code is documented (docstrings, project documentation).
1. [x] GraphQL schema and type definitions are up to date.
1. [ ] Changes are mentioned in the changelog.

### CODE REVIEW

File: saleor/graphql/core/utils/reordering.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    to an absolute sorting.
+
+    This will then commit the changes onto the nodes.
+
+    :param qs: The query set on which we want to retrieve and reorder the node.
+    :param operations: The operations to make: {pk_to_move: +/- 123}.
+    :param field: The field from which nodes are resolved.
+
+    :raises RuntimeError: If the bulk operation is not run inside an atomic transaction.
+    :raises ValueError: If a primary key is invalid. Will raise ValidationError if
```
Comment (NyanKiyoshi) [inline saleor/graphql/core/utils/reordering.py]: No longer true

File: saleor/graphql/core/utils/reordering.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    pk: int
+    sort_order: int
+
+
+class Reordering:
+    def __init__(
+        self, qs: QuerySet, model_name: str, operations: Dict[int, int], field: str
+    ):
+
+        if not transaction.get_connection().in_atomic_block:
```
Comment (NyanKiyoshi) [inline saleor/graphql/core/utils/reordering.py]: should be moved out of the class (inside `perform_reordering`)

**Comment by django-queries** [comment]
Here is the report for 59635cad73c21970cd10ac45705b0585244136ba (NyanKiyoshi/saleor @ sort_order/shifting)
Base comparison is 1e248962a0689101677d294cb0d98a67228a9742.



<details><summary>No differences were found. (click me)</summary>
<p>

```diff
# api.benchmark checkout
  test name                           	left count 	right count	duplicate count
  ------------------------------------	-----------	-----------	---------------
  add billing address to checkout     	         34	         34	             20
  add shipping to checkout            	          7	          7	              0
  checkout payment charge             	         14	         14	              0
  complete checkout                   	          6	          6	              0
  create checkout                     	         43	         43	             21

# api.benchmark homepage
  test name                           	left count 	right count	duplicate count
  ------------------------------------	-----------	-----------	---------------
  retrieve main menu                  	          5	          5	              0
  retrieve product list               	          4	          4	              0
  retrieve secondary menu             	          5	          5	              0
  retrieve shop                       	          2	          2	              0

# api.benchmark product
  test name                           	left count 	right count	duplicate count
  ------------------------------------	-----------	-----------	---------------
  product details                     	         15	         15	              3

# api.benchmark variant
  test name                           	left count 	right count	duplicate count
  ------------------------------------	-----------	-----------	---------------
  retrieve variant list               	         15	         15	              6
```

</p>
</details>

File: saleor/graphql/schema.graphql [hunk 3]
```diff
@@ -2322,7 +2322,7 @@ input ProductFilterInput {
 }
 
 type ProductImage implements Node {
-  sortOrder: Int!
+  sortOrder: Int
```
[the remark this answers is not in the document]
Comment (maarcingebala) [reply saleor/graphql/schema.graphql]: Same for `MenuItem` type.

File: saleor/graphql/core/utils/reordering.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        return node_pos, target_pos, target_position
+
+    def process_move_operation(self, pk, move):
+        old_sort_order = self.ordered_node_map[pk]
+
+        try:
+            node_pos, target_pos, new_sort_order = self.calculate_new_sort_order(
+                pk, move
+            )
+        except NoChangeException:
```
Comment (maarcingebala) [inline saleor/graphql/core/utils/reordering.py]: This is fine but I think we could simplify the code if we had something like this:

```
def process_move_operation(self, pk, move):
    if move == 0:
        warnings.warn(...)
        return

    self.calculate_new_sort_order(...)
    # continue with sorting logic
```

This way we're not introducing additional concept such as a custom exception, which is used only in this file.

File: saleor/graphql/schema.graphql (hunk 3 above)
[the remark this answers is not in the document]
Comment (NyanKiyoshi) [reply saleor/graphql/schema.graphql]: It would be better to wait for #4353 before dropping it

**Pull request merged**